One-Click Installer

straussmaximilian opened the pull request:
Hi,
this is to have a one-click installable version of OmicLearn.

## Example for Mac:

### Installer: 
<img width="122" alt="Screen Shot 2022-05-24 at 7 47 32 PM" src="https://user-images.githubusercontent.com/16521288/170099749-39ec22cf-ff86-4322-bc9d-6a30d149a801.png">

### App:
<img width="100" alt="Screen Shot 2022-05-24 at 7 49 16 PM" src="https://user-images.githubusercontent.com/16521288/170100023-c8dd8311-398e-4b78-9bd7-070b372709c8.png">

## Changes
This required me to restructure everything a bit and create a python project structure. I also removed the docker-compose and the conda environment as we are not using docker anymore and everything can be installed via pip.

I already compiled some test-installers on my [fork](https://github.com/straussmaximilian/OmicLearn/releases/tag/t003):

@Furkan: Could you do the following?
- [x] Check the changes
- [ ] Run the new release action
- [ ] Test the installers in case the action runs through

I tested them briefly locally and seemed to work. Note that for mac you will need to right click open to trust the app and for Windows you might need admin rights etc.

As the next steps toward resubmission wanted to do the following:
- [ ] Update the long manuscript version with the one-click functionality
- [ ] Write some installation instructions for the one-click installers
After that, we can also update the wiki:
- [ ] Update Wiki

furkanmtorun commented on requirements_pip.txt:
```diff
@@ -6,4 +6,5 @@ openpyxl==3.0.9
-watchdog==2.1.7
\ No newline at end of file
+watchdog==2.1.7
+xgboost
```
Should we also pin the xgboost version here?

furkanmtorun commented on omiclearn/__init__.py:
```diff
@@ -0,0 +1,38 @@
+__license__ = "Apache"
+__description__ = "Transparent exploration of machine learning for biomarker discovery from proteomics and omics data"
+__author__ = "OmicEra"
+__author_email__ = "omiclearn@omicera.com"
```
Do we know that this account still exists? Maybe, we might simply change it to "info@omicera.com".

furkanmtorun reviewed (changes requested):
Hi Max,

Thanks for your all efforts.
I have left some comments for some points, and here are some major ones:

1) I tested the one-click installation on my MacAir. It worked for me and sounds fancy but the installation took really long time (over 20 minutes).  Same for the opening up the application took a long time. How was it on your Mac?
(nit): We could also play around with the logo issue (top-left corner):

![Screenshot 2022-05-25 at 01 25 46](https://user-images.githubusercontent.com/49681382/170145051-5b16f243-42f1-449b-8886-5228f1bdd5a1.png)


2) As for Docker, we might only keep the Dockerfile for other users who want to deploy on their own machines.


Regarding the next steps,

1) As prev. discussed, we can use the long version and replace the "one-click" installation with "Docker" if we resolve the aforementioned issues and then, we can update the Wiki!

2) We still have a "Session history" problem due to my current implementation as mentioned in my prev. email and I could not figure it out with Streamlit's built-in "Session state" functionality. Overall, in the local installation, it works expectedly but not on Streamlit hosting weirdly:

![Screenshot 2022-05-25 at 01 53 53](https://user-images.githubusercontent.com/49681382/170145033-02e78c77-08e8-4664-be14-20fe4835b6c5.png)

Best,
Furkan

straussmaximilian commented:
Hi Furkan,

Thanks for the quick reply.

1) For me, the installation on my M1 Mac took maybe 5 minutes, not the fastest, but 20 min would be too long. I will investigate what this could be and will check the logo position. Did you check it for Windows / Linux? If Mac is too slow, we can still go for the other versions.

2) We can go for docker - however you had removed it in the past:
(https://github.com/OmicEra/OmicLearn/commit/74c82921f186cb70450e1d569da7086ca2e18b69). I assumed this was intended to not use docker, I hence removed the `docker-compose` as this will not work w/o the dockerfile.  I noticed there is also [renku](https://renkulab.io), which we could then use as an "academic" alternative to the streamlit hosting.

For the streamlit hosting bug with the session state, I can investigate: Could you tell me what you tried / what didn't work and if you have any clue where it could come from.

On this note I also noticed that there seems to be an issue with the font color, e.g., the descriptions are not visible. This could explain why it might be confusing for some people to use. Do you have any idea how to fix this?

Thanks!

<img width="733" alt="Screen Shot 2022-05-25 at 10 26 21 AM" src="https://user-images.githubusercontent.com/16521288/170217189-ef66376f-2321-48d7-98b1-3a7b49caf0e2.png">

<img width="727" alt="Screen Shot 2022-05-25 at 10 26 30 AM" src="https://user-images.githubusercontent.com/16521288/170217176-202a1c17-8f3e-4579-b02b-e888d0df674b.png">

furkanmtorun commented:
Dear Max,

1. Thanks for checking out in advance! Now, I tested installation on a Windows machine, and only took 1 minute to install and worked perfectly! As for the slow installation issue, it might be specific to my Mac as it is slow itself. 

2. Ah, actually I thought I had remained the Dockerfile itself in the repo but now, I realized that it is docker-compose :) My fault. Then, let's go without Docker. Sorry for the confusion again.

3. Actually, I do not have a full understanding of the reason why it fails in Streamlit hosted version but works in the local version. Previously, it was working for both versions however with the current implementation of [`utils/session_states.py`](https://github.com/OmicEra/OmicLearn/blob/master/utils/session_states.py), it only works for local installations. 
Then, checked [Streamlit built-in "Session state" function](https://docs.streamlit.io/library/api-reference/session-state), however, could not integrate and make it work with `save_sessions()` in `utils/ui_helper.py` and [line 513 at `omic_learn.py`](https://github.com/OmicEra/OmicLearn/blob/3867fa8765156da501adb12c8359fede3153ea9f/omic_learn.py#L513).

4. Great catch! Thanks for letting me know! To color those labels inside the expanding boxes, we could insert the following line of CSS into `utils/ui_helper.py`:
```.streamlit-expanderContent .css-qrbaxs {color: #035672 !important}```

   It will only affect the elements having class `.css-qrbaxs` inside `.streamlit-expanderContent` boxes not the others that share the same class name.

Have a great day!

straussmaximilian commented:
Hi Furkan,

I made another update. The session state and the font color should now be fixed. I also updated the Readme and WIki. I decided to remove the TOC as it was not up-to-date in some cases, and the TOC is also shown in the GitHub sidebar.

I used the markdown files and converted them to PDF with this [tool](https://www.npmjs.com/package/md-to-pdf); `md-to-pdf ./**/*.md`. We can then use this as an additional Supplementary Information. [wiki_compiled.pdf](https://github.com/OmicEra/OmicLearn/files/8824922/wiki_compiled.pdf)

Could you check the code and merge it? Then run the release action and test the installer?

For merging: It looks like the CLA Assistant is broken, so either someone with the respective rights can fix this or someone merges with admin privileges.

I will update the manuscript next.

furkanmtorun reviewed (approved):
Dear Max,

Thanks for your updates and I apologize for my late reply due to participation in a conference.

- New wiki looks nice I would say! Great job, thanks a lot!
- Session state still works in local & hopefully will work in the Streamlit cloud when we deploy
- I also really liked the idea of compiled Wiki pages in one pdf file as `OmicLearn Manual` to provide as Supp. Material as you said. 

(nit) For readme file and Wiki installation page, could you please change this line:
- Install OmicLearn with `pip install .`
to
- Install OmicLearn with `cd OmicLearn & pip install .`

As for merging, I can ask someone who has admin rights for helping us to merge the PR!

Have a nice day!
